RFC.md

```diff
@@ -1,12 +1,234 @@
+to understand the multidimensional examples as given above, and we
+are forced to make conservative assumptions, inhibiting optimisation. 
+This information must ideally be expressible in LLVM, such that LLVM's optimisers and
+alias analysis can use this information to model multidimensional-array
+semantics.
+
+
+
+## Evaluation of the impact of the intrinsic on accuracy of dependence analysis
+
+- This has been implemented in an exprimental branch of Polly, and was used 
+on the COSMO climate weather model. This greatly helped increase the accuracy
+of Polly's analysis, since we eliminated the guessing game from the array analysis.
+
+- This has also been implemented as part of a GSoC effort to unify
+Chapel and Polly. 
+
+- Molly, the distributed version of Polly written by Michael Kruse for his PhD
+  also implemented a similar scheme. In his use-case, guessing as we perform
+  optimistically was not possible, so this kind of intrinsic was _necessary_
+  for the application, not just _good to have_. More details are available
+  in his PhD thesis: [Lattice QCD Optimization and Polytopic Representations of Distributed Memory](https://tel.archives-ouvertes.fr/tel-01078440).
+  In particular, Chapter 9 contains a detailed discussion.
+
+# Representations
+
+## Instruction / Intrinsic
+
+### Syntax
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> {<stride>, <idx>}*
+```
+
+### Overview:
+The `multidim_array_index` instruction is used to get the address of 
+an element from an array. It performs address calcuation only and 
+does not access memory. It is similar to `getelementptr`. However, it imposes
+additional semantics which allows the optimiser to provide better optimisations
+than `getlementptr`.
+
+
+
+### Arguments:
+The first argument is always a type used as the basis for the calculations. The
+second argument is always a pointer, and is the base address to start the
+calculation from. The remaining arguments are a list of pairs. Each pair
+contains a dimension stride, and an offset with respect to that stride.
+
+
+### Semantics:
+
+`multidim_array_index` represents a multi-dimensional array index, In particular, this will
+mean that we will assume that all indeces `<idx_n>` are non-negative.
+
+
+##### Address computation:
+Consider an invocation of `multidim_array_index`:
+
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> <str_0>, <idx_0>, <str_1> <idx_1>, ..., <str_n> <idx_n>
+```
+
+If the pairs are denoted by `(str_i, idx_i)`, where `str_i` denotes the stride
+and `idx_i` denotes the index of the ith pair, then the final address (in bytes)
+is computed as:
+
+```
+addressof(ptrval) + len(ty) * [(str_0 * idx_0) + (str_1 * idx_1) + ... (str_n * idx_n)]
+```
+
+## Transitioning to `multidim_array_index`: Allow `multidim_array_index` to refer to a GEP instruction:
+
+```llvm
+i1 foo(... arr):
+    %gep = getelementptr ...
+    %multidim = multidim_array_array(%size0, %size1, %ix0, %ix1, %arr, %gep)
```

Comment:
This does not match the order of arguments defined.  Are you trying to define another variant of multidim_array_index? Why does the name become `multidim_array_array` ? Should the title be changed to be consistent?

Is the idea here that the front-ends emit a GEP to compute the index "the old way" and also a multidim_array_index to compute it the "new way" and then optimizations unaware of multidim_array_index can just look at the GEP? If so, wouldn't we want the original instruction proposal above to have a place to put the GEP?

Reply:
Ah yes, this was one of the choices that I was considering, but don't think it makes much sense now since it adds an extra parameter which we will have to move away from in any case...